core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -0,0 +1,471 @@
+          }
+
+          String entityTag = Hex.encodeHexString(digest);
+          return Response.ok().tag(entityTag).build();
+        } catch (Exception e) {
+          throw toObjectS3Exception(e, objectPath);
+        }
+      }
+    });
+  }
+
+  /**
+   * @summary retrieves an object's metadata
+   * @param bucket the bucket name
+   * @param object the object name
+   * @return the response object
+   */
+  @HEAD
+  @Path(OBJECT_PARAM)
+  @ReturnType("java.lang.Void")
+  public Response getObjectMetadata(@PathParam("bucket") final String bucket,
+      @PathParam("object") final String object) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
+      @Override
+      public Response call() throws S3Exception {
+        Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
+        Preconditions.checkNotNull(object, "required 'object' parameter is missing");
+
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        AlluxioURI objectURI = new AlluxioURI(objectPath);
+
+        try {
+          URIStatus status = mFileSystem.getStatus(objectURI);
+          // TODO(cc): Consider how to respond with the object's ETag.
+          return Response.ok()
+              .lastModified(new Date(status.getLastModificationTimeMs()))
+              .header(S3Constants.S3_CONTENT_LENGTH_HEADER, status.getLength())
+              .build();
+        } catch (Exception e) {
+          throw toObjectS3Exception(e, objectPath);
+        }
+      }
+    });
+  }
+
+  /**
+   * @summary downloads an object
+   * @param bucket the bucket name
+   * @param object the object name
+   * @return the response object
+   */
+  @GET
+  @Path(OBJECT_PARAM)
+  @ReturnType("java.io.InputStream")
+  @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_OCTET_STREAM})
+  public Response getObject(@PathParam("bucket") final String bucket,
+      @PathParam("object") final String object) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
+      @Override
+      public Response call() throws S3Exception {
+        Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
+        Preconditions.checkNotNull(object, "required 'object' parameter is missing");
+
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        AlluxioURI objectURI = new AlluxioURI(objectPath);
+
+        try {
+          URIStatus status = mFileSystem.getStatus(objectURI);
+          FileInStream is = mFileSystem.openFile(objectURI);
+          // TODO(cc): Consider how to respond with the object's ETag.
+          return Response.ok(is)
```

Comment:
as this is repeated in HEAD and GET, does it make sense to have a utility for URIStatus to Response conversion?

Reply:
it's slightly different where GET has `is` as the entity. We don't get much code simplicity because we would have to return a `ResponseBuilder` in the helper function